pkg/client/client_test.go

```diff
@@ -1910,7 +1946,303 @@ func TestClose(t *testing.T) {
+	s.client.OnTrack(onTrack)
+	err := s.client.Track("sample_conversion", userContext, map[string]interface{}{})
+
+	s.NoError(err)
+	s.False(isTrackCalled)
+}
+
+func (s *ClientTestSuiteTrackEvent) TestTrackNotificationNotCalledWhenInvalidParamsProvided() {
+
+	isTrackCalled := false
+	onTrack := func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent) {
+		isTrackCalled = true
+	}
+
+	s.client.OnTrack(onTrack)
+	err := s.client.Track("bob", entities.UserContext{ID: "1212121", Attributes: map[string]interface{}{}}, map[string]interface{}{})
+
+	s.NoError(err)
+	s.Equal(0, len(s.mockProcessor.Events))
+	s.False(isTrackCalled)
+}
+
+func (s *ClientTestSuiteTrackEvent) TestTrackNotificationNotCalledWhenSendThrowsError() {
+
+	s.mockProcessor.On("ProcessEvent", mock.Anything).Return(true)
+	expectedUserContext := entities.UserContext{ID: "1212121", Attributes: map[string]interface{}{}}
+
+	isTrackCalled := false
+	onTrack := func(eventKey string, userContext entities.UserContext, eventTags map[string]interface{}, conversionEvent event.ConversionEvent) {
+		isTrackCalled = true
+	}
+
+	mockNotificationCenter := new(MockNotificationCenter)
+	config, err := s.client.GetProjectConfig()
+	s.NoError(err)
+	configEvent, err := config.GetEventByKey("sample_conversion")
+	s.NoError(err)
+	userEvent := event.CreateConversionUserEvent(config, configEvent, expectedUserContext, map[string]interface{}{})
+	expectedTrackNotification := notification.TrackNotification{EventKey: "sample_conversion", UserContext: expectedUserContext, EventTags: map[string]interface{}{}, ConversionEvent: *userEvent.Conversion}
+
+	mockNotificationCenter.On("Send", notification.Track, expectedTrackNotification).Return(fmt.Errorf(""))
+	mockNotificationCenter.On("AddHandler", notification.Track, mock.AnythingOfType("func(interface {})")).Return(1, nil)
+	s.client.NotificationCenter = mockNotificationCenter
+	s.client.OnTrack(onTrack)
+	err = s.client.Track("sample_conversion", expectedUserContext, map[string]interface{}{})
+
+	s.NoError(err)
+	s.Equal(1, len(s.mockProcessor.Events))
+	s.False(isTrackCalled)
+	mockNotificationCenter.AssertExpectations(s.T())
+}
+
+type ClientTestSuiteTrackNotification struct {
+	suite.Suite
+	mockProcessor       *MockProcessor
+	mockDecisionService *MockDecisionService
+	client              OptimizelyClient
+}
+
+func (s *ClientTestSuiteTrackNotification) SetupTest() {
+	mockProcessor := new(MockProcessor)
+	mockProcessor.On("ProcessEvent", mock.Anything).Return(true)
+	s.mockProcessor = mockProcessor
+	s.mockDecisionService = new(MockDecisionService)
+	s.client = OptimizelyClient{
+		ConfigManager:      ValidProjectConfigManager(),
+		DecisionService:    s.mockDecisionService,
+		EventProcessor:     s.mockProcessor,
+		NotificationCenter: notification.NewNotificationCenter(),
+	}
+}
+
+func (s *ClientTestSuiteTrackNotification) TestAddandRemoveOnTrack() {
```

Comment:
Please break down this test into 2 or 3 tests. 
- Add multiple ontrack and assert
- Add multiple ontrack, remove by ID and assert then. 
- Add, remove, add and assert. 
Make sure, IDs are always increased(unique).